luigi/contrib/bigquery_avro.py

```diff
@@ -0,0 +1,78 @@
+"""Specialized tasks for handling Avro data in BigQuery from GCS.
+"""
+import avro
+from itertools import iter, next
+import logging
+
+from luigi.contrib.bigquery import BigQueryLoadTask, SourceFormat
+from luigi.contrib.gcs import GCSClient
+from luigi.task import flatten
+
+
+logger = logging.getLogger(__name__)
+
+
+class BigQueryLoadAvro(BigQueryLoadTask):
+    """A helper for loading specifically Avro data into BigQuery from GCS.
+
+    Additional goodies - takes field documentation from the input data and propagates it
+    to BigQuery table description and field descriptions.
+
+    Suitable for use via subclassing: override requires() to return Task(s) that output
+    to GCS Targets; their paths are expected to be URIs of .avro files or URI prefixes
+    (GCS "directories") containing one or many .avro files.
+
+    Override output() to return a BigQueryTarget representing the destination table.
+    """
+    source_format = SourceFormat.AVRO
+
+    def source_uris(self):
+        return [x.path + '/*.avro' for x in flatten(self.input())]
+
+    def _get_input_schema(self):
+        '''Arbitrarily picks an object in input and reads the Avro schema from it.'''
+        input_target = next(iter(flatten(self.input())))
+        input_fs = input_target.fs if hasattr(input_target, 'fs') else GCSClient()
+        input_uri = next(iter(self.source_uris()))
+        if '*' in input_uri:
+            input_uri = next(iter(input_fs.list_wildcard(input_uri)))
+
+        schema = []
+
+        def read_schema(fp):
+            # We rely on that the DataFileReader will initialize itself fine as soon as the file
+            # header with schema is downloaded, without requiring the remainder of the file...
+            try:
+                reader = avro.datafile.DataFileReader(fp, avro.io.DatumReader())
+                schema.append(reader.datum_reader.writers_schema)
+            except Exception as e:
+                logger.info('%s', e)
+                return False
+            return True
+
+        input_fs.download(input_uri, 64 * 1024, read_schema).close()  # TODO check with various chunksizes (suspect the file position might matter)
+
+        return schema[0]
```

Comment:
Yeah, a crash inside download will occur if no Avro files exist (or if the first of source_uris() doesn't point to an existing file). Will see if the error message from there is too cryptic. Perhaps it's self explanatory.